src/main/scala/org/allenai/pipeline/s3/S3Artifact.scala

```diff
@@ -122,45 +115,110 @@ trait S3Artifact[A <: Artifact] extends Logging {
     service.putObject(request)
   }
 
-  protected var cachedFile: Option[A] = None
+  def readContents(): InputStream = service.getObject(bucket, path).getObjectContent
+
+  def download(file: File, bufferSize: Int = 1048576): Unit = {
+    logger.debug(s"Downloading $bucket/$path to $file")
+    file.deleteOnExit()
+    val os = new FileOutputStream(file)
+    val is = readContents()
+    val buffer = new Array[Byte](bufferSize)
+    Iterator.continually(is.read(buffer)).takeWhile(_ != -1).foreach(n =>
+      os.write(buffer, 0, n))
+    is.close()
+    os.close()
+  }
+}
 
-  private val BUFFER_SIZE = 1048576
-  // 1 MB buffer
-  protected def getCachedArtifact: A = cachedFile match {
-    case Some(f) => f
-    case None =>
-      val cacheDir = {
-        val f = new File(System.getProperty("java.io.tmpdir"), "pipeline-cache")
-        if (f.exists && !f.isDirectory) {
-          f.delete()
-        }
-        if (!f.exists) {
-          f.mkdirs()
-        }
-        f
+/** Implements policy for making local caches of artifacts in S3
+  */
+trait S3Cache {
+  def readFlat(artifact: S3Artifact): InputStream
+
+  def readZip(artifact: S3Artifact): Reader
+
+  def writeFlat[T](artifact: S3Artifact, writer: ArtifactStreamWriter => T): T
+
+  def writeZip[T](artifact: S3Artifact, writer: Writer => T): T
+}
+
+object DefaultS3Cache extends LocalS3Cache()
+
+/** Caches S3 artifacts on the local filesystem
+  * @param persistentCacheDir If specified, cached objects will be stored in a directory
+  *                           that is persistent across JVM processes. There are no limits
+  *                           on the size of the data stored in the persistent directory
+  */
+case class LocalS3Cache(
+    persistentCacheDir: Option[File] = Some(new File(System.getProperty("java.io.tmpdir"), "pipeline-cache"))
+) extends S3Cache with Logging {
+
+  override def readFlat(artifact: S3Artifact) = {
+    if (usePersistentCache) {
+      new FileArtifact(downloadLocalCopy(artifact)).read
+    } else {
+      artifact.readContents()
+    }
+  }
+
+  def usePersistentCache = persistentCacheDir.isDefined
+
+  override def writeFlat[T](artifact: S3Artifact, writer: ArtifactStreamWriter => T): T = {
+    val local = createLocalCopy(artifact)
+    val result = new FileArtifact(local).write(writer)
+    artifact.upload(local)
+    result
+  }
+
+  override def readZip(artifact: S3Artifact): Reader = {
+    new ZipFileArtifact(downloadLocalCopy(artifact)).reader
+  }
+
+  override def writeZip[T](artifact: S3Artifact, writer: Writer => T) = {
+    val local = createLocalCopy(artifact)
```

Comment:
Zip files still have to be downloaded locally, but will be deleted on JVM exit if no persistent cache directory is specified.